modules/storage-page-memory/src/main/java/org/apache/ignite/internal/storage/pagememory/mv/BlobStorage.java

```diff
@@ -0,0 +1,329 @@
+
+        super.reuseList = reuseList;
+        this.statisticsHolder = statisticsHolder;
+    }
+
+    /**
+     * Reads a blob stored starting at a page with the given ID.
+     *
+     * @param firstPageId ID of first page.
+     * @return Byte array for the blob.
+     * @throws IgniteInternalCheckedException If something goes wrong.
+     */
+    public byte[] readBlob(long firstPageId) throws IgniteInternalCheckedException {
+        ReadState readState = new ReadState();
+
+        long pageId = firstPageId;
+
+        while (pageId != NO_PAGE_ID) {
+            Boolean ok = PageHandler.readPage(
+                    pageMem,
+                    grpId,
+                    pageId,
+                    PageLockListenerNoOp.INSTANCE,
+                    readFragment,
+                    readState,
+                    0,
+                    false,
+                    IoStatisticsHolderNoOp.INSTANCE
+            );
+
+            assert ok : pageId;
+
+            pageId = readState.nextPageId;
+        }
+
+        assert readState.bytes != null;
+
+        return readState.bytes;
+    }
+
+    /**
+     * Adds a new blob to the storage.
+     *
+     * @param bytes Blob bytes.
+     * @return ID of the page starting the chain representing the blob.
+     * @throws IgniteInternalCheckedException If something goes wrong.
+     */
+    public long addBlob(byte[] bytes) throws IgniteInternalCheckedException {
+        return doStore(NO_PAGE_ID, bytes);
+    }
+
+    /**
+     * Updates the blob content.
+     *
+     * @param firstPageId ID of the first page in the chain storing the blob.
+     * @param bytes New blob content.
+     * @throws IgniteInternalCheckedException If something goes wrong.
+     */
+    public void updateBlob(long firstPageId, byte[] bytes) throws IgniteInternalCheckedException {
+        doStore(firstPageId, bytes);
+    }
+
+    private long doStore(long maybeFirstPageId, byte[] bytes) throws IgniteInternalCheckedException {
+        Objects.requireNonNull(bytes, "bytes is null");
+
+        long firstPageId = allocatePageIfNeeded(maybeFirstPageId);
+
+        WriteState state = new WriteState(bytes);
+        state.pageId = firstPageId;
+
+        do {
+            Boolean ok = PageHandler.writePage(
+                    pageMem,
+                    grpId,
+                    state.pageId,
+                    PageLockListenerNoOp.INSTANCE,
+                    writeFragment,
+                    null,
+                    state,
+                    0,
+                    false,
+                    statisticsHolder
+            );
+
+            assert ok : state.pageId;
+        } while (!state.stop);
+
+        freePagesStartingWith(state.firstPageToFreeId);
+
+        return firstPageId;
+    }
+
+    private long allocatePageIfNeeded(long maybePageId) throws IgniteInternalCheckedException {
```

Comment:
I think you can use `org.apache.ignite.internal.pagememory.datastructure.DataStructure#allocatePage(org.apache.ignite.internal.pagememory.reuse.ReuseBag)` with `null`

Reply:
This is what I do

Reply:
Ok, `allocatePage()` I do not think that it is needed, it is used in one place.

Reply:
Inlined the method